data/src/main/java/io/gnosis/data/repositories/TransactionRepository.kt

```diff
@@ -253,6 +258,38 @@ class TransactionRepository(
         }
 }
 
+private fun TransferDto.isIncoming(safeInfo: SafeInfo): Boolean {
+    if (to == safeInfo.address && from != safeInfo.address) {
+        return true
+    }
+    return false
+}
+
+private fun EthereumTransactionDto.isIncoming(safeInfo: SafeInfo): Boolean {
+    if (to == safeInfo.address && from != safeInfo.address) {
+        return true
+    }
+    return false
+}
+
+private fun MultisigTransactionDto.isIncoming(safeInfo: SafeInfo): Boolean {
+    if (dataDecoded?.method == "transfer") {
+        return false
+    } else if (dataDecoded?.method == "transferFrom" || dataDecoded?.method == "safeTransferFrom") {
+        val from = dataDecoded.parameters?.getValueByName("from")?.asEthereumAddress()
+        val to = dataDecoded.parameters?.getValueByName("to")?.asEthereumAddress()
+
+        if (from != safeInfo.address && to == safeInfo.address) {
+            return true
+        }
+        return false
+    } else {
+        // unknown method
+        return false
+    }
+}
+
+
```

Comment:
All these methods would be better suited IMO in a mapping that happens closer to the UI. Here we are really coupling transaction requests with safe requests. It will hit us with composability of these objects. Besides, `Transactions.Transfer` should already contain `sender` and `recipient`. If there is a bug there, then the problem is in the mapping, we are maybe assigning the wrong address to the wrong field.

Reply:
If we want to calculate the incoming status in the UI layer then we need to pass up the dataDecoded (method, from and to ) section. Is that really worth it?

Reply:
Unfortunately sender & recipient do not determine alone whether a transfer is incoming or outgoing. Here is the full description: https://docs.google.com/document/d/1vhTu5NbB-6m8nE3cqTS3bRONigPO3NA5YkJqNTA6AOk/edit?skip_itp2_check=true